src/nimble_block_material_interface_base.h

```diff
@@ -0,0 +1,85 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                                NimbleSM
+//                             Copyright 2018
+//   National Technology & Engineering Solutions of Sandia, LLC (NTESS)
+//
+// Under the terms of Contract DE-NA0003525 with NTESS, the U.S. Government
+// retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
+// NO EVENT SHALL NTESS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions?  Contact David Littlewood (djlittl@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef SRC_NIMBLE_BLOCK_MATERIAL_INTERFACE_BASE_H
+#define SRC_NIMBLE_BLOCK_MATERIAL_INTERFACE_BASE_H
+
+namespace nimble {
+
+class BlockBase;
+class Material;
+
+struct BlockData {
+  BlockData(nimble::BlockBase *block_, nimble::Material* material_d_, const int block_id_,
+            const int num_block_elems_, const int num_points_per_block_elem_)
+      :
+      block(block_),
+      material_device(material_d_),
+      id(block_id_),
+      num_elems(num_block_elems_),
+      num_points_per_elem(num_points_per_block_elem_) {
+  }
+
+  nimble::BlockBase *block = nullptr;
```

Comment:
Storing a pointer to `BlockBase` is the tricky part here, since there is still not a clean interface to `NGPLameData`, `dynamic_cast`s to `nimble_kokkos::Block` are required in Extras. This can be part of a future refactor if needed.

Reply:
@japlews
I noticed that `std::shared_ptr<nimble::NGPLAMEData> ngp_lame_data_` is a member of `nimble_kokkos::Block` as a _placeholder_ (?). 

What do you think if I put this member in the base class and add to the base class interface the routine `std::shared_ptr<nimble::NGPLAMEData> GetNGPLAMEData()` ? 

If that helps, I could do the similar modification for `nimble_kokkos::MaterialFactory` and `nimble::MaterialFactoryBase`

Reply:
@uhetmaniuk I vote to leave it as is—more of just a reminder for me that this awkward relationship should be fixed. It needs to be moved out of NimbleSM anyway; it's the last holdover from inverting the dependency of NimbleSM -> Extras.

Reply:
@japlews OK. Will keep it as discussed. Thank you.